akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala

```diff
@@ -96,23 +59,20 @@ private[akka] class InputStreamPublisher(is: InputStream, bytesReadPromise: Prom
     readBytes match {
       case -1 ⇒
         // had nothing to read into this chunk
-        eofReachedAtOffset = readBytes
-        log.debug("No more bytes available to read (got `-1` or `0` from `read`), marking final bytes of file @ " + eofReachedAtOffset)
+        log.debug("No more bytes available to read (got `-1` from `read`)")
+        onCompleteThenStop()
 
       case _ ⇒
         readBytesTotal += readBytes
-        if (readBytes == chunkSize) availableChunks :+= ByteString1C(arr)
-        else availableChunks :+= ByteString1C(arr).take(readBytes)
 
-      // valid read, continue
+        // emit immediately, as this is the only chance to do it before we might block again
+        onNext(ByteString1C(arr).take(readBytes))
```

Comment:
Isn’t it cheaper to keep using the same array of chunkSize for reading and copying into an appropriately-sized one before emitting? Especially for workloads where this change matters readBytes will frequently be significantly less than chunkSize, wasting a lot of space.